### DISCUSSION TITLE
scrollback documentation for 22J doesn't seem to work properly

### DISCUSSION THREAD

**Discussion by llimllib** [kovidgoyal/kitty#6460] [Q&A]
So the new `22J` escape code will save my scrollback is exactly what I want.

If I run this in my terminal, I get exactly the behavior I want:

```
printf "\e[H\e[22J"
```

My scollback gets saved, terminal gets cleared, and my prompt gets shown normally.

However, if I follow the documentation's instructions for installing it to my `~/.zshrc`:

```shell
clear-screen-saving-contents-in-scrollback() {
    printf "\e[H\e[22J"
}

zle -N clear-screen-saving-contents-in-scrollback
bindkey '^l' clear-screen-saving-contents-in-scrollback
```

My scollback gets saved, my terminal gets cleared, but I'm left without a PS1 prompt.

Anybody know how I can properly bind `^l` in zsh such that it works exactly like the printf?

**Comment by kovidgoyal** [comment]
https://github.com/kovidgoyal/kitty/commit/4a039a45ec34c40cf999dacc28ed0696c3b79f68

**Reply by llimllib** [reply]
Thank you! that's an improvement, but it still adds an extra empty prompt to the scrollback, which I'd prefer to avoid. Do you know of any way to avoid that?

<img width="603" alt="image" src="https://github.com/kovidgoyal/kitty/assets/7150/d713aef5-64de-43d9-a334-07b9102fc795">

the sequence for that test was:

- run a command with a lot of output
- `^l`, to clear the screen, bound to the command you posted above
- `cmd-f` (bound to `map cmd+f launch --stdin-source=@screen_scrollback --stdin-add-formatting --type=overlay less +G -R` in kitty pref)

**Comment by llimllib** [comment]
I tried a bunch of different zle commands to add after the `printf`. None were perfect, but at least they improve the situation by putting me back at a prompt:

```
clear-screen-saving-contents-in-scrollback() {
    printf "\e[H\e[22J"
    # these work, but add an extra line to the scrollback:
    # zle reset-prompt
    zle redisplay
    # zle clear-screen
    #
    # this works, but leaves a blank line at the top:
    # zle accept-line
    # these don't work:
    # zle kill-line
    # zle end-of-line
    # zle get-line
}
```

**Comment by kovidgoyal** [comment]
On Thu, Jul 13, 2023 at 09:25:41AM -0700, Bill Mill wrote:
> Thank you! that's an improvement, but it still adds an extra empty prompt to the scrollback, which I'd prefer to avoid. Do you know of any way to avoid that?

Nope, I am not a zsh expert you might be better off asking in a zsh
forum.

**Reply by kovidgoyal** [reply]
Actually this will work at least for single line prompts
https://github.com/kovidgoyal/kitty/commit/d4f9d22c920b387fc2cd5813ede3434e570ff8c6

for muti line prompts you will need to figure out how to move the cursor to the top prompt line using some zsh function or the other.

**Reply by llimllib** [reply]
dang, that's so close! I tried a bunch of stuff but haven't been successful yet. Thanks for looking at it
